File: services/rag/src/embeddings/cohere.ts
import { withRetry } from "../utils/retry.js";
import type { EmbeddingClient, EmbeddingConfig } from "./types.js";

export class CohereEmbedding implements EmbeddingClient {
  private readonly apiKey: string;
  readonly model: string;
  readonly dimensions: number;
  private readonly baseUrl = "https://api.cohere.com/v2";

  // Cohere's max texts per request
  private static readonly MAX_BATCH = 96;

  constructor(config: EmbeddingConfig) {
    this.apiKey = config.apiKey;
    this.model = config.model || "embed-multilingual-v3.0";
    this.dimensions = config.dimensions ?? 1024;
  }

  async embed(texts: string[]): Promise<number[][]> {
    if (texts.length > CohereEmbedding.MAX_BATCH) {
      const results: number[][] = [];
      for (let i = 0; i < texts.length; i += CohereEmbedding.MAX_BATCH) {
        const batch = texts.slice(i, i + CohereEmbedding.MAX_BATCH);
        const batchResults = await this.embed(batch);
        results.push(...batchResults);
      }
      return results;
    }

    return withRetry(async () => {
      const response = await fetch(`${this.baseUrl}/embed`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${this.apiKey}`,
        },
        body: JSON.stringify({
          model: this.model,
          texts,
          input_type: "search_document",
          embedding_types: ["float"],
        }),
      });

      if (!response.ok) {
        const error = await response.text();
        throw new Error(`Cohere embedding failed (${response.status}): ${error}`);
      }

      const data = (await response.json()) as {
        embeddings: { float: number[][] };
      };
      return data.embeddings.float;
    });
  }

  async embedSingle(text: string): Promise<number[]> {
    const [embedding] = await this.embed([text]);
    return embedding;
  }
}

File: services/confluence-importer/src/confluence_importer/embeddings/cohere.py
"""Cohere embeddings via the REST API (no SDK dependency)."""

import httpx

from confluence_importer.embeddings.base import EmbeddingApiError, batched, validate_dimensions
from confluence_importer.retry import with_retry

_BASE_URL = "https://api.cohere.com/v2"
_MAX_BATCH = 96


class CohereEmbedding:
    def __init__(
        self,
        *,
        api_key: str,
        model: str,
        dimensions: int = 1024,
        http_client: httpx.Client | None = None,
    ) -> None:
        self.model = model or "embed-multilingual-v3.0"
        self.dimensions = dimensions
        self._http = http_client or httpx.Client(
            headers={"Authorization": f"Bearer {api_key}"},
            timeout=60.0,
        )

    def embed_texts(self, texts: list[str]) -> list[list[float]]:
        results: list[list[float]] = []
        for batch in batched(texts, _MAX_BATCH):
            results.extend(self._embed_batch(batch))
        return results

    def embed_query(self, text: str) -> list[float]:
        return self.embed_texts([text])[0]

    def _embed_batch(self, texts: list[str]) -> list[list[float]]:
        def do_request() -> list[list[float]]:
            response = self._http.post(
                f"{_BASE_URL}/embed",
                json={
                    "model": self.model,
                    "texts": texts,
                    "input_type": "search_document",
                    "embedding_types": ["float"],
                },
            )
            if response.status_code >= 400:
                raise EmbeddingApiError("Cohere", response.status_code, response.text)
            embeddings = response.json()["embeddings"]["float"]
            return validate_dimensions("Cohere", self.model, self.dimensions, embeddings)

        return with_retry(do_request)
